### DISCUSSION TITLE
关于下一步开发什么功能

### DISCUSSION THREAD

**Discussion by jc3213** [jc3213/download_with_aria2#52] [Ideas]
1） 多用户档案——思路来自 https://github.com/jc3213/download_with_aria2/discussions/22 有个雏形，但是现在底层改了不少内容，要直接适配过来貌似并不容易？最大的问题是，火狐在`downloads` API的工作模式很不合理，Chromium我倒是想到了可以多档案都拥有自己的设置。而且也就查询文件扩展的时候会慢点，但是火狐不更改API的前提下基本确定会先下载文件再取消发送给jsonrpc，最后增加磁盘IO显得非常愚蠢。但是我没经历两个浏览器做不同的适配

2） 快速抓取按钮——思路来自 https://github.com/jc3213/download_with_aria2/discussions/33 ，总体来说本意为添加一个功能方便快速把当前页面添加到抓取目录里面。但是不管是做成右键菜单还是popup菜单，或者是快捷键，都缺乏标识……要写标识又感觉太复杂。目前暂时没有好的方案

3） 下载所有图像——测试用版本
[sniffer_all_images_v5.2.zip](https://github.com/jc3213/download_with_aria2/files/10707071/sniffer_all_images_v5.2.zip)

**Comment by ivysrono** [comment]
快速抓取按钮如果集成起来很不方便，有没有考虑做成独立的扩展？保持主扩展的独立和简洁，更多功能作为周边来提供。
类似于浏览器和扩展的关系。

**Reply by jc3213** [reply]
这个就更加得知识范围外了。虽然理论上通过message就能实现。主要还是怎么做出已经存在于列表中，点击取消，同时还涉及多个规则这种。UI这里我直接就搞不定了。在上个讨论就已经做了个雏形但是完全无解……

**Reply by ivysrono** [reply]
这就是我选择那个b站批量下载脚本的原因了~

**Reply by jc3213** [reply]
UI太麻烦了，虽然现在2年过去了进步了不少就是了，我现在做的这个也只是勉强能用，反正头皮发麻。我最开始搞开发的时候直接用javascript做的UI

**Reply by jc3213** [reply]
对了1654别推送火狐版本，我单独弄出来是为了丢给M$审核，刚发现了个错误会破坏右键菜单。我去标注一下得了

**Comment by ivysrono** [comment]
不记得之前有没有提过，大概没有：能否做个默认忽略IP地址的选项。
IP地址的网站一般都是内网，没法适配的，速度又快也没必要使用Aria2，所以统统禁掉比较方便。

**Reply by jc3213** [reply]
怎么说呢，有些网站的下载地址也是IP，比如当年的惠普驱动啥的储存的服务器（还是戴尔来着），暴力匹配IP并不可取。各个公司内网的IP规则我也不清楚。反正匹配这里我本身就不在行，匹配方法都是按照原扩展进行优化的来的。

**Reply by jc3213** [reply]
要么改成纯正则，慢是慢点，但是理论上能直接输入 192.168.1.* 这种泛匹配规则。就是不知道一个超长正则匹配起来性能如何了……

当然，这个对于Chromium版本的影响微乎其微，因为发生的位置比火狐版本超前，反而火狐版本会受到严重的性能影响，导致更容易出现2次下载（火狐开始下载 => 火狐取消下载 => aria2开始下载 => aria2完成下载）

**Reply by jc3213** [reply]
换成了includes而不是endsWith，这样你可以使用 192.168.1. 去匹配 192.168.1.0~255，更新了[Wiki: Match Pattern](//github.com/jc3213/download_with_aria2/wiki/MatchPattern)

**Comment by jc3213** [comment]
试着做了下新版的多档案适配，貌似因为`保存`跟`撤销`还有`恢复`这3个按钮的原因导致非常复杂，这里必须单独适配。感觉优先级直接变成最低了。虽然储存方面的方案倒是想好了

**Comment by jc3213** [comment]


**Reply by ivysrono** [reply]
好好休息。身体不适就不要想着用爱发电了。

**Reply by jc3213** [reply]
身边好像都没我家这么久，一家人都在咳嗽。而且我家我最严重，咳得最厉害。唉。但是相比最开始那几天好很多了，但就是一个超级慢性的过程。

**Reply by ivysrono** [reply]
个体差异非常非常大。

**Reply by jc3213** [reply]
反正我现在看到所谓的专家讲新冠症状，怎么应对就想骂人，我这边医院发热门诊全部排队根本没床位，而且肺部出现炎症全靠激素跟抗生素处理，对老年人来说完全就是死马当活马医。对年轻人来说这种肺部损伤+大剂量激素下去身体也会废一半。

关键这病毒还会伤害心肌啥的，福建那边朴朴超市因为要员工退烧就去上班结果有人送货路上心肌炎发作病亡。家属去拉横幅要求给说法。而所谓的XX医生在网络上各种说不可怕，怎么应对，然后推荐特效药。事实就是新冠根本没有特效药，辉瑞那个也只是安慰剂。就算真正有效也只是感染前5天。问题是你出现症状的时候已经是3天以后了。普遍会错过最佳服药时机，而且3000/盒宰人。

然后YY专家就在电视上跟主流媒体上忽悠大家别担心，反正我是听说我这边的殡仪馆跟火葬场已经排队了。然后新冠死亡统计还不算你有基础病，或者死在家里的。当年嘲笑印度，现在国内这情况真的不好笑。我这边因为提前爆发，很多人听信专家的3~6个月内不会感染，不少小孩已经不戴口罩了（虽然现在寒假）

**Reply by ivysrono** [reply]
无论何时，戴好口罩！

**Reply by jc3213** [reply]
症状每天好转，但是每天睡眠都不好，倒下咳嗽，然后精神就好了（至少大脑清醒）每天折腾半天才睡，昨天鼻子还堵搞得严重没睡好。反正等过段时间好转了以后我考虑把设置重新弄一下。

**Reply by daiaji** [reply]
咳嗽貌似最长可能是8周，反正喉咙不舒服咳出黄痰就是有感染，用抗生素，然后用一些化痰的药，比方说盐酸氨溴索。

**Reply by jc3213** [reply]
我是白痰，所以也没用抗生素，化痰的药是真不想吃。能不吃药不吃药。反正就是不完全好，每天好一丢丢，挺恶心的。现在大量的人不戴口罩。反正我现在贼怕二次感染。

**Reply by daiaji** [reply]
还是吃点化痰的药吧，上呼吸道粘膜损坏引起的应激还是蛮影响心情的。

**Reply by jc3213** [reply]
反正就跟平时感冒后咽炎发作差不多，也就那样了吧。反正现在各路专家还在忽悠。就难受。对了，下面提到的新分支有机会帮忙看看吧。目前只是把简单功能做出来但是后续不知道怎么搞。

**Comment by jc3213** [comment]
@ivysrono @daiaji @wordpure 帮忙测试看看 [新分支：newtabmanager](https://github.com/jc3213/download_with_aria2/tree/newtabmanager)，看看有啥建议或者意见没。尤其是弹窗模式下《以标签页打开任务管理器》这个按钮相关看看怎么说？目前我弄了个雪花图标，感觉有点蠢

**Reply by jc3213** [reply]
更换了下思维，现在mv2的两个分支完全没问题了（popup那边还会优化就是了）。配合选项已经能正常切换点击工具栏菜单后是直接弹出还是新标签页打开。

但是mv3……我得慢慢折腾。感觉就很折磨

**Reply by ivysrono** [reply]
能创新，有成果，看来你脑雾好差不多了

**Reply by jc3213** [reply]
我没脑雾，就是咳嗽，之前单纯是睡眠不好或者年前没精力。最近稍微有空了——其实是因为怕死不想出去聚餐，一个人在家里呆着图个清静。

顺带一提，🐏康以后变得懒惰表示想躺平（主要还是期间各种专家忽悠搞得心态炸了摆烂了）

**Reply by ivysrono** [reply]
放羊为主，摸鱼为辅，放过自己，珍爱生命

**Reply by jc3213** [reply]
目前就是这个mv3激活方面问题不小，整体来说目前的方案也不是不能接受。实在不行每个`addListener`都写一遍获取数据，谁怕谁。f@ck

最大的问题是chrome大量API都是交互形式存在时差，但是`addListener`又是激活便执行。mv2之前还好，mv3以后因为后台`service worker`会休眠，你全局储存的数据会抹除。导致如果`addListener`里面有获取储存的代码都必须重新获取一遍。

谷歌的说法是这样更节能，更安全，毕竟权限没了，不能常驻后台了，但给我感觉是更傻逼。明明IO变多了。

**Reply by jc3213** [reply]
另外这里也是影响档案开发的罪魁祸首，同样的原因导致唤醒后因为没有了全部全局数据，必须重新获取，但是获取这个交互的过程中，`contextMenus.onClicked`事件已经执行到后面需要用到储存信息的地方，直接报错。反正这个mv3，真的蠢得要死

**Reply by jc3213** [reply]
功能装上了，也顺便优化了下底层逻辑。就是任务管理器跟新下载还有待后续优化。关于如何识别已存在的任务管理器这里我也会重新做下逻辑。不过那也是年后的事情了。

总之，新年快乐

**Comment by jc3213** [comment]
话说能帮忙找一个禁止盗链的图库？或者说有没有方法从cache里面读取图片的缓存并显示到页面里的API？感觉迷了

**Reply by ivysrono** [reply]
https://addons.mozilla.org/firefox/addon/anti-anti-leech/

**Reply by jc3213** [reply]
我想测试下新功能而已。并不是要扩展，OTL一时间想不到啥图库是禁止外链的

**Reply by ivysrono** [reply]
我是想说这扩展里的规则是明文的，readme里有很多测试例。

**Reply by jc3213** [reply]
目前看来交互逻辑有问题。从background打开newpopup，无法确认其是否完全load，因此接着发出sendMessage，popup无法接收到。必须setTimeout延迟，或者类似new download那样由popup发起，然后background回应。

但是这里就涉及到background => content => background， 打开popup后，popup => background => popup。这种通信请求。必须通过2个message变量去搞定。我现在麻了

**Reply by jc3213** [reply]
而且还涉及到MV3这个新的按钮怎么i18n。反正头皮发麻

**Reply by daiaji** [reply]
~Referer 似乎不允许修改了。~
~不知道在做什么妖。~
https://tiebapic.baidu.com/forum/pic/item/dadf721273f0820288b91bf90efbfbedaa641b67.jpg?tbpicau=2023-02-11-05_294cb5065f7556c2a80d3617a3bc3f58
艳红搞了个更阴间的解决方案。
图片URL后面有个ID。
去掉就会……

**Reply by jc3213** [reply]
我并不是想用类似扩展去逃过图库的检测。而是想试试通过自己的方案实现本地预览图像URL。毕竟我下一个功能想做下载所有图像（然后由用户自己选定）。现在碰到的就是交互问题，UI布局，然后就是MV3在service worker不能使用chrome.i18n的大前提下，怎么实现这个菜单的i18n。反正就觉得挺搞的。

**Reply by jc3213** [reply]
目前做了个雏形，交互的URL都是指定死了的。而且是由background延迟向popup发起请求。但是头皮发麻感觉太多东西要改。反正我麻了。

**Reply by daiaji** [reply]
还是提一嘴，叔叔搞了Referer请求白名单，Referer为贴吧就能请求图片，Referer为知乎就会403.
还真是不要妨碍叔叔挣钱。
虽说和你的目的无关。

**Reply by jc3213** [reply]
B站？还是百度？能不能给个图片链接我试试

**Reply by jc3213** [reply]
> ~Referer 似乎不允许修改了。~ ~不知道在做什么妖。~ https://tiebapic.baidu.com/forum/pic/item/dadf721273f0820288b91bf90efbfbedaa641b67.jpg?tbpicau=2023-02-11-05_294cb5065f7556c2a80d3617a3bc3f58 艳红搞了个更阴间的解决方案。 图片URL后面有个ID。 去掉就会……

我测试了下，扩展内部的窗口直接引用可以正常显示，直接点击打开反而403。那么问题不大。我大概可以做UI了。刚顺手优化了下交互逻辑。目前交互应该没问题了。

**Reply by jc3213** [reply]
目前v2已经可以测试了，其实除了多了个context menu，而且这个menu只能打开一个新的popup，然后显示你这个百度图片外，没别的东西了。抓取全部images的逻辑没做，生成UI的逻辑没做，UI交互没做。只是做了个简单的框架

**Reply by jc3213** [reply]
@ivysrono 帮忙测试一下新版。我这边测试问题不大。基本操作就是右键下载所有图片（目前英文菜单），然后弹出窗口，根据预览选择图片。选择了的图潘会变暗，然后点`确认`后就下载了。

目前还没启用下载，只是做console.log。如果你确定要使用下载功能，必须修改`/res/js/sniffer.js`把17行的`//`给删除。

![sample](https://user-images.githubusercontent.com/8744936/217982466-c1c0e4fd-106a-4a08-b794-2a3e5ed723eb.png)

**Reply by jc3213** [reply]
v4调整了UI大小，给图片增加了border，毕竟部分图片可见部分很小，但是实际面积很大。

**Reply by jc3213** [reply]
测试工作正常后发布v4.1，但是目前还无法传递任何参数啥的。麻烦帮忙测试一下。另外UI方面有撒建议吗？

**Reply by jc3213** [reply]
才发现4.1火狐跟mv3的background没改，现在v5改好了。目前还在测试中

**Reply by ivysrono** [reply]
争取周末找个时间看下

**Reply by jc3213** [reply]
话说YAWW我发现使用`http://token:mytoken@localhost:6800/jsonrpc`无法正常工作提示未授权。这个是我在某个electron应用中发现的。不知道是不是我自己的问题。不管是这个electron内部的YAWW，或者外部的YAWW，都只要有`--rpc-secret=mytoken`这种选项，直接就裂开。而我的扩展，`Aria2 WebUI`反而没问题，如果没记错我开始开发之前用的`Aria2Ng`好像也正常

**Reply by daiaji** [reply]
> B站？还是百度？能不能给个图片链接我试试

叔叔和艳红都在搞。
策略不同而已。
测了，浏览历史好像没同步过来。
豆瓣好像也在搞吧。

**Reply by jc3213** [reply]
我只想知道下载所有图片这个功能是否正常弹出图片。目前没做css里面的background-image:('url:yyyyy')的适配，并且重复的图片链接都并不会多次显示

**Reply by daiaji** [reply]
我压根没时间测试，在痛苦适配coreboot中。

**Reply by jc3213** [reply]
惨……好吧，我目前还被YAAW卡住了。反正挺垃的。这个web ui

**Reply by daiaji** [reply]
适配BIOS，玩是真好玩，难顶也是真难顶。

**Reply by jc3213** [reply]
这个YAAW，某个百度网盘软件内置的，无法识别token导致授权错误，直接用webui，提示内部服务器错误。AriaNG，Aria2 WebUI屁事没有。

我就没搞懂为什么对着url一定要用`http://user:pass@host.com:6800/jsonrpc`这种格式，而不是把secret单独做个选项放出来，老的user, pass早就不推荐使用了。为什么一定要用正则表达式去匹配。直接用`new URL()`不是更好。官方的还没那么多纰漏。

我现在用这个网盘的软件的时候，毕竟自己也会启动一个完全没有token的aria2c进程，但是会对我这个有token的无限发起认证请求导致CPU占用上升。

反正我这边把这破东西拉黑了。真的烂

**Comment by jc3213** [comment]
在目前的工作逻辑下面，适配多档案参数必须改成强制弹出`新下载窗口`，这样能不怎么修改底层。两个页面稍微做个适配就行了，额外增加一个`multiprofile.js`就好了

**Comment by jc3213** [comment]
发现特定网站，比如百度网盘还有别的一些，非常吃user-agent，因此考虑重写user-agent部分。改成可编辑的。另外，我还准备调整几个策略组的UI。这个貌似比较重要。我先研究下。
